homeassistant/components/speedtestdotnet/__init__.py

```diff
@@ -41,45 +40,130 @@
 
-    if not conf[CONF_MANUAL]:
-        async_track_time_interval(hass, data.update, conf[CONF_SCAN_INTERVAL])
+    client = SpeedTestClient(hass, config_entry)
+    hass.data[DOMAIN] = client
+
+    if not await client.async_setup():
+        return False
 
-    def update(call=None):
-        """Service call to manually update the data."""
-        data.update()
+    return True
 
-    hass.services.async_register(DOMAIN, "speedtest", update)
 
-    hass.async_create_task(
-        async_load_platform(
-            hass, SENSOR_DOMAIN, DOMAIN, conf[CONF_MONITORED_CONDITIONS], config
-        )
-    )
+async def async_unload_entry(hass, config_entry):
+    """Unload SpeedTest Entry from config_entry."""
+    hass.services.async_remove(DOMAIN, SPEED_TEST_SERVICE)
+    if hass.data[DOMAIN].unsub_timer:
+        hass.data[DOMAIN].unsub_timer()
+
+    await hass.config_entries.async_forward_entry_unload(config_entry, "sensor")
+
+    hass.data.pop(DOMAIN)
 
     return True
 
 
-class SpeedtestData:
+class SpeedTestClient:
     """Get the latest data from speedtest.net."""
 
-    def __init__(self, hass, server_id):
+    def __init__(self, hass, config_entry):
         """Initialize the data object."""
+        self.hass = hass
+        self.config_entry = config_entry
         self.data = None
-        self._hass = hass
-        self._servers = [] if server_id is None else [server_id]
+        self.api = None
+        self.unsub_timer = None
 
-    def update(self, now=None):
+    def update(self):
         """Get the latest data from speedtest.net."""
 
         _LOGGER.debug("Executing speedtest.net speed test")
-        speed = speedtest.Speedtest()
-        speed.get_servers(self._servers)
-        speed.get_best_server()
-        speed.download()
-        speed.upload()
-        self.data = speed.results.dict()
-        dispatcher_send(self._hass, DATA_UPDATED)
+
+        if self.config_entry.options[CONF_SERVER_ID]:
+            server_id = [self.config_entry.options[CONF_SERVER_ID]]
+            self.api.get_servers(servers=server_id)
+            self.api.closest.clear()
+
+        self.api.get_best_server()
+        self.api.download()
+        self.api.upload()
+        self.data = self.api.results.dict()
+
+        async_dispatcher_send(self.hass, DATA_UPDATED)
+        _LOGGER.debug("Speed test data updated")
+
+    async def async_setup(self):
+        """Set up SpeedTest."""
+        self.api = speedtest.Speedtest()
+        self.add_options()
+        self.set_scan_interval(self.config_entry.options)
+
+        async def update(service=None):
+            """Service call to manually update the data."""
+            await self.hass.async_add_executor_job(self.update)
+
+        self.hass.services.async_register(DOMAIN, SPEED_TEST_SERVICE, update)
+
+        self.hass.async_create_task(
+            self.hass.config_entries.async_forward_entry_setup(
+                self.config_entry, "sensor"
```

Comment:
"sensor" should be coming from the constant as it was before.

Reply:
This wasn't used before because this is forwarding the config_entry to the sensor platform.

Reply:
Import was on line 8 and the use was on line 59.

Reply:
Oh ok, sure we can import the const.